### PR TITLE
Adding support for chef_gem wihch is available for newer versions of chef

### PR DESCRIPTION
**Pull request by emourant**
Adding forwards compatibility support for installing "mongo" gem via chef_gem. The cookbook could sometimes error with later versions of chef as the gem was not being installed correctly.

### CODE REVIEW

File: recipes/default.rb [hunk 1]
```diff
@@ -23,14 +23,21 @@
...
+needs_mongo_gem = (node.recipes.include?("mongodb::replicaset") or node.recipes.include?("mongodb::mongos"))
 
+# install the mongo ruby gem at compile time to make it globally available
 if needs_mongo_gem
-  # install the mongo ruby gem at compile time to make it globally available
-  gem_package 'mongo' do
-    action :nothing
-  end.run_action(:install)
-  Gem.clear_paths
+  current_version = Gem::Version.new(Chef::VERSION)
```
Comment (jeremyolliver) [inline recipes/default.rb:30]: :+1: for this overall Pull Request to prefer usage of `chef_gem` when available, and for the approach of checking the chef version to see if it's available.

However I'd like to note that `Gem::Version.new` is only available in ruby 1.9, which although this is packaged with the current omnibus installer into /opt/chef/, older chef client installations may not be using an isolated ruby, and could have ruby 1.8 - perhaps a straight string compare against Chef::VERSION might be more compatible in that regard?

**Comment by msaffitz** [comment]
:+1: just spent 20 minutes chasing this issue, thinking it was a conflict with rbenv

**Comment by josephholsten** [comment]
Is this compatible with Amazon OpsWare's chef 0.9?

**Comment by jamesonjlee** [comment]
I don't think we should support OpsWorks chef 0.9  We really shouldn't be expected to support chef versions that have been out of support for year(s).  OpsWorks you can set the default ruby to 1.9.3 now so this looks like it should be fine.

**Comment by iceydee** [comment]
Why isn't this getting merged? I also spent a long time chasing the same issue.

**Comment by jamesonjlee** [comment]
@iceydee i've held back on merging this in because of this requiring ruby >= 1.9.  I am currently (forced) to use ruby 1.8.7.
 @emourant could you make the changes to make this PR support ruby 1.8.7?

**Comment by jeremyolliver** [comment]
Just checking back on this, mostly because I suspect my original comment about requiring ruby 1.9 may be incorrect, I think I was confused about another issue I'd recently encountered with Gem::Version not supporting four part version numbers until rubygems 1.8.0. Technically, it's the rubygems version which provides this method, and rubygems can be upgraded independent of the ruby version. Also, I just ran a quick test, and on a local ruby 1.8.7-p352 install which has the quite old rubygems 1.6.2, this works fine.

TL;DR;
- I think this _does_ actually work on 1.8.7
- Opsworks now supports chef 11

I don't see any reason not to merge this

**Comment by josephholsten** [comment]
If memory serves, old rubygems shipped with the same class implementation, just in a different namespace.

This is annoying to look up on a phone because rubygems wasn't technically part of the stdlib in 1.8. Looks like latest 1.8.x rubygems also uses Gem::version https://github.com/rubygems/rubygems/blob/master/lib/rubygems/version.rb

## Any idea what release of rubygems we need to support?

http://josephholsten.com

**Comment by jeremyolliver** [comment]
The oldest version on the github rubygems mirror (7 years back) is https://github.com/rubygems/rubygems/blob/8e0965a5e12752864ba8377163a01c82d06d16e2/lib/rubygems/version.rb and does _visually_ appear to have `Gem::Version` in there with an initialize and comparator, though I do remember that `Gem::Dependency` used to be the common interface.

I've tested on an older chef 0.10.X ubuntu 10.0.4 setup which uses ruby via the standard ubuntu apt package, which is ruby 1.8 and rubygems 1.6.2 - This supports `Gem::Version` fine.

Given that "doesn't support ruby 1.8" is obviously wrong, I'd say test out the following on the opsworks 0.9 platform you want to support (I can't find any documentation on what version it runs).

```
# irb
irb> require 'rubygems'
irb> v = Gem::Version.new("1.0.0")
```

If that works, full steam ahead.

**Comment by iceydee** [comment]
Could also check for it?

Like:

```
  Gem.const_defined?("Version")
```

If that returns false, we can assume that it doesn't support the chef_gem resource and fallback on gem_package.
